msticpy/nbtools/timeline.py

```diff
@@ -0,0 +1,940 @@
+
+
+# pylint: disable=too-many-locals
+def _display_timeline_grouped(
+    data: pd.DataFrame, **kwargs
+) -> figure:  # noqa: C901, MC0001
+    """
+    Display a timeline of events.
+
+    Parameters
+    ----------
+    data : pd.DataFrame
+        Input DataFrame
+
+    Other Parameters
+    ----------------
+    group_by : str
+        The column to group timelines on
+    sort_by : str
+        (where `data` is a DataFrame)
+        The column to order timelines on
+    ref_time : datetime, optional
+        Input reference line to display (the default is None)
+    title : str, optional
+        Title to display (the default is None)
+    time_column : str, optional
+        Name of the timestamp column
+        (the default is 'TimeGenerated')
+    legend_column : str, optional
+        Name of the column used to generate the legend lables
+        (the default is None/no legend)
+    legend: str, optional
+        Where to position the legend
+        None, left, right or inline (default is None)
+    yaxis : bool, optional
+        Whether to show the yaxis and labels
+    source_columns : list
+        List of source columns to use in tooltips
+        (the default is None)
+    height : int, optional
+        The height of the plot figure
+        (the default is auto-calculated height)
+    width : int, optional
+        The width of the plot figure (the default is 900)
+    color : str
+        Default series color (default is "navy")
+
+    Returns
+    -------
+    figure
+        The bokeh plot figure.
+
+    """
+    reset_output()
+    output_notebook()
+
+    height: int = kwargs.pop("height", None)
+    width: int = kwargs.pop("width", 900)
+    ref_time: Any = kwargs.pop("ref_time", None)
+    title: str = kwargs.pop("title", None)
+    time_column: str = kwargs.get("time_column", "TimeGenerated")
+    source_columns: list = kwargs.pop("source_columns", None)
+    group_by: str = kwargs.get("group_by", None)
+    sort_by: str = kwargs.pop("sort_by", None)
+    show_yaxis: bool = kwargs.pop("yaxis", False)
+    show_range: bool = kwargs.pop("range_tool", True)
+    color: str = kwargs.get("color", "navy")
+
+    graph_df, group_count_df, tool_tip_columns, series_count = _create_data_grouping(
+        data, source_columns, time_column, group_by, sort_by, color
+    )
+
+    # build the tool tips from columns (excluding these)
+    tool_tip_items = [(f"{col}", f"@{col}") for col in tool_tip_columns]
+    hover = HoverTool(tooltips=tool_tip_items, formatters={"Tooltip": "printf"})
+
+    title = f"Timeline: {title}" if title else "Event Timeline"
+    min_time = graph_df[time_column].min()
+    max_time = graph_df[time_column].max()
+    start_range = min_time - ((max_time - min_time) * 0.1)
+    end_range = max_time + ((max_time - min_time) * 0.1)
+    height = height if height else _calc_auto_plot_height(series_count)
+
+    plot = figure(
+        x_range=(start_range, end_range),
```

Comment:
We probably also want to set y_range here.

Reply:
Agreed - +/- 25% of min/max